src/syntax/grammar.mly

```diff
@@ -42,6 +42,11 @@ let rec psep sep f = parser
 	| [< r = psep_nonempty sep f >] -> r
 	| [< >] -> []
 
+let rec psep_trailing sep f = parser
+	| [< v = f; '(sep2,_) when sep2 = sep; l = psep_trailing sep f >] -> v :: l
+	| [< v = f >] -> [v]
```

Comment:
This case is useless because the first pattern is the same as in the case above. Doesn't this give a warning?

Reply:
How is it useless? The first case matches a sequence including a separator, this one matches even without a separator.

Reply:
Huh, I had it committed to memory that left factorization is not supported in camlp4 stream parsers, but apparently I'm wrong about that.

I also didn't realize that `when` clauses are allowed on subsequent patterns. This seems odd because these patterns are expected ones. How would this behave if `f` is matched but `sep` isn't:

```ml
let rec psep_trailing sep f = parser
	| [< v = f; '(sep2,_) when sep2 = sep; l = psep_trailing sep f >] -> v :: l
	| [< >] -> []
```

Does this error because it expects `sep` or does it do something else?

Reply:
[Left factorisation](https://camlp5.github.io/doc/pdf/camlp5-7.12.pdf#subsection.4.4.2) is apparently done for consecutive rules by camlp5. This might also answer your question about the `when` on the second element. I don't have an easy way to check with OCaml right now though.